{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "mcb111_section_w01.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1Vm7j-4Op7qC"
      },
      "source": [
        "# Information content of DNA sequences\n",
        "\n",
        "Often in the field of bioinformatics a question comes to mind: **What is the shortest (motif) that we expect to appear only once?** or **What is the shortest length that can be uniquely found in the genome?** \n",
        "\n",
        "This is important to find out when you are trying to figure out if a DNA sequence of interest is appearing in a genome simply by chance or it is a significant appearance.\n",
        "\n",
        "Eukaryotic genomes have on the order of L=$10^6−10^{10}$ nucleotides. We can define $l$, an arbitrary length of a motif of interest. We can also begin by saying that the composition of different bases (A, C, G, T) is balanced such that we have equal amounts of each throughout the genome. \n",
        "\n",
        "Let's generate a function that generates genomes with predetermined characteristics, so we can test some of these concept empirically:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "gjjyjXFPrSsr"
      },
      "source": [
        "def generate_genome(size, g_content, c_content, a_content):\n",
        "\n",
        "  ''' \n",
        "  This function generates a genome of user defined size and GC content.\n",
        "\n",
        "  Parameters:\n",
        "    - Number of bases.\n",
        "    - % of GC content.\n",
        "  '''\n",
        "\n",
        "  g_comp = ['G']*g_content\n",
        "  c_comp = ['C']*c_content\n",
        "  a_comp = ['A']*a_content\n",
        "  t_comp = ['T']*(100-(g_content+c_content+a_content))\n",
        "\n",
        "  human_genome_composition = np.array(g_comp+a_comp+t_comp+c_comp) # making array with target composition\n",
        "  human_genome = np.random.choice(human_genome_composition, int(size)) # sampling from composition to generate full genome.\n",
        "\n",
        "  return ''.join(human_genome) # returns dull genome string form"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ny020EUwZGd8",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "324e2730-4ee2-4caa-8700-b1d829ec7eaa"
      },
      "source": [
        "%pylab inline\n",
        "from IPython.display import clear_output"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Populating the interactive namespace from numpy and matplotlib\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4iZv3EWgzfQL",
        "outputId": "fe1d5bd6-c546-4b08-eaef-cbfed356d164"
      },
      "source": [
        " %%time\n",
        " genome = generate_genome(size = 10e6, g_content = 25, c_content = 25, a_content=25)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CPU times: user 2.48 s, sys: 788 ms, total: 3.27 s\n",
            "Wall time: 3.28 s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "zVWCMbEGQ6XZ",
        "outputId": "fc0626ca-254e-42c4-ad2f-f69793c7b436"
      },
      "source": [
        "genome[:20]"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'AATCGGACATAAGCCGACTA'"
            ]
          },
          "metadata": {},
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ilJ6tguazf5_",
        "outputId": "95bfe6e0-9ff2-488b-dee1-dced93d8ea8b"
      },
      "source": [
        "bases , counts =  np.unique(list(genome), return_counts=True)\n",
        "comps = counts/10e6*100\n",
        "\n",
        "print(\"Composition of our mock genome:\\n\"\n",
        "        + bases[0]+\": \" + str(round(comps[0],2)) + \" %\\n\"\n",
        "        + bases[1]+\": \" + str(round(comps[1],2)) + \" %\\n\"\n",
        "        + bases[2]+\": \" + str(round(comps[2],2)) + \" %\\n\"\n",
        "        + bases[3]+\": \" + str(round(comps[3],2)) + \" %\" )"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Composition of our mock genome:\n",
            "A: 24.98 %\n",
            "C: 25.03 %\n",
            "G: 25.0 %\n",
            "T: 24.99 %\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Fk1f_wXbJSlG"
      },
      "source": [
        "Now we can play with our genome. For instance if we pick our motif to be “A”, it is expected to land in 1/4 of the positions in the genome. ~2.5M times"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uUzgZ4Zq3TGb",
        "outputId": "f43e7e43-dbfc-4886-89f3-3caca1b4e24d"
      },
      "source": [
        "genome.count('A')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "2498436"
            ]
          },
          "metadata": {},
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "692eOw4RJjyJ"
      },
      "source": [
        "If we take a motif of length l=2, for instance “AA” it will match fewer than 1/4, but not much better. ~0.5M times"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Oi9CmMaqRXCx"
      },
      "source": [
        "s = 'AAAA'"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "J6KhZ9iORYPW",
        "outputId": "81061e75-6b41-45e2-aa46-b5fb66056a11"
      },
      "source": [
        "s.count('AA')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "2"
            ]
          },
          "metadata": {},
          "execution_count": 12
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "c5ziLBGJKRUv",
        "outputId": "cce70598-08c1-4521-85d3-4ef7d24ffb05"
      },
      "source": [
        "genome.count('AA')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "499303"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "R_xdXLwGeWpP"
      },
      "source": [
        "We kindof see a pattern here, the longer the sequence the less it is likely to be found in the genome. Lets try to build an intuition for the information content of a sequence by following a series of thought experiments that we can then validate through code.\n",
        "\n",
        "First Let's imagine we have a genome with 50% G, 50% C and 0% A or T composition when we sample from it sequentially, we know that we are only going to get Gs and Cs\n",
        "\n",
        "How **surprised** would we be if the next sample that we took was either a G or a C? not much, right? Because these are the only two possibilitties in this example. If we **expect** a G or a C with a 50% chance we would be equally surprised if we got one or the other.\n",
        "\n",
        "However, it is hard to say how **surprised** we would be (infinitely?) if the next sample was an A or a T. So it really isn't well defined but that's ok because we are talking about events that can't happen."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 279
        },
        "id": "mc0i0L2mg3G0",
        "outputId": "3aa00088-4040-4802-8153-8f6248280524"
      },
      "source": [
        "genome = generate_genome(size = 1000, g_content = 50, c_content = 50, a_content=0) #Make genome size manageable for dynamic plotting.\n",
        "As = 0\n",
        "Ts = 0\n",
        "Gs = 0\n",
        "Cs = 0\n",
        "\n",
        "while (As+Ts+Gs+Cs)<=200:\n",
        "    \n",
        "    clear_output(wait=True)\n",
        "    fig,ax = subplots(ncols= 1,nrows =1, sharey = False)\n",
        "    ax.bar(['A','C','T','G'],[As,Cs,Ts,Gs])\n",
        "    ax.set_xlabel('Nucleotide')\n",
        "    ax.set_ylabel('Count')\n",
        "    sample = np.random.choice(list(genome),1) # making the string genome a list to sample easily\n",
        "\n",
        "    if sample == 'C':\n",
        "        Cs += 1\n",
        "    if sample == 'T':\n",
        "        Ts += 1\n",
        "    if sample == 'G':\n",
        "        Gs += 1\n",
        "    if sample == 'A':\n",
        "        As += 1        \n",
        "    plt.show()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEGCAYAAACKB4k+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAPyUlEQVR4nO3dfbBcdX3H8fcHAiJF5elOGgENVqpFVNSrgthWhT98aAUtUhhGI8Vm2lofqwXbP7Cd6RRaH2sdmRSV0KEKIhasDkp5snYUvVAUgaIpEg0T4Kr4bItpv/1jT36uMSHLzd09udn3a2Zn9/x+5+z5Zmcnn3t+Z8/vpKqQJAlgt74LkCTtPAwFSVJjKEiSGkNBktQYCpKkZlnfBeyIAw88sFauXNl3GZK0pNxwww3fqqqZrfUt6VBYuXIlc3NzfZchSUtKkvXb6nP4SJLUGAqSpMZQkCQ1hoIkqTEUJEmNoSBJagwFSVJjKEiSGkNBktQs6SuaJU2vlWd+ou8SenXn2S8ay/t6pCBJagwFSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSY2hIElqDAVJUmMoSJIaQ0GS1BgKkqRmbKGQ5ANJ7k3ylaG2/ZNcmeRr3fN+XXuS/F2SdUm+nOSp46pLkrRt4zxSOB94/hZtZwJXVdVhwFXdMsALgMO6x2rgfWOsS5K0DWMLhar6DPCdLZqPB9Z2r9cCJwy1X1ADnwf2TbJiXLVJkrZu0ucUllfVxu713cDy7vVBwDeH1tvQtf2CJKuTzCWZm5+fH1+lkjSFejvRXFUF1AK2W1NVs1U1OzMzM4bKJGl6TToU7tk8LNQ939u13wUcMrTewV2bJGmCJh0KlwOrutergMuG2l/R/QrpKOB7Q8NMkqQJWTauN07yIeA5wIFJNgBnAWcDFyc5HVgPnNSt/knghcA64MfAaeOqS5K0bWMLhao6ZRtdx25l3QJePa5aJEmj8YpmSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSY2hIElqDAVJUmMoSJIaQ0GS1Ixt7iPt2lae+Ym+S+jVnWe/qO8SpLHwSEGS1BgKkqTGUJAkNYaCJKkxFCRJjaEgSWoMBUlSYyhIkhpDQZLUGAqSpMZQkCQ1hoIkqTEUJEmNoSBJagwFSVJjKEiSGkNBktT0EgpJ3pDkliRfSfKhJHslOTTJ9UnWJbkoyZ591CZJ02zioZDkIOC1wGxVHQHsDpwMnAO8s6oeC9wHnD7p2iRp2vU1fLQMeGiSZcDewEbgecAlXf9a4ISeapOkqTXxUKiqu4C3Ad9gEAbfA24AvltVm7rVNgAHbW37JKuTzCWZm5+fn0TJkjQ1+hg+2g84HjgUeCTwS8DzR92+qtZU1WxVzc7MzIypSkmaTn0MHx0HfL2q5qvqp8ClwDHAvt1wEsDBwF091CZJU62PUPgGcFSSvZMEOBa4FbgGOLFbZxVwWQ+1SdJU6+OcwvUMTijfCNzc1bAGOAN4Y5J1wAHA+yddmyRNu2XbX2XxVdVZwFlbNN8BPKOHciRJHa9oliQ1hoIkqTEUJEmNoSBJagwFSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSY2hIElqDAVJUmMoSJIaQ0GS1BgKkqTGUJAkNYaCJKkxFCRJjaEgSWoMBUlSYyhIkhpDQZLUGAqSpMZQkCQ1I4VCkmNGaZMkLW2jHim8Z8Q2SdIStuyBOpMcDTwLmEnyxqGuhwO7j7MwSdLkPWAoAHsC+3TrPWyo/fvAieMqSpLUjwcMhaq6DrguyflVtX6xdppkX+A84AiggN8DbgcuAlYCdwInVdV9i7VPSdL2jXpO4SFJ1iT5dJKrNz92YL/vBq6oqscDTwZuA84Erqqqw4CrumVJ0gRtb/hos48A5zL46/5/d2SHSR4B/AbwSoCquh+4P8nxwHO61dYC1wJn7Mi+JEkPzqihsKmq3rdI+zwUmAc+mOTJwA3A64DlVbWxW+duYPnWNk6yGlgN8KhHPWqRSpIkwejDRx9P8kdJViTZf/NjgftcBjwVeF9VPQX4EVsMFVVVMTjX8Auqak1VzVbV7MzMzAJLkCRtzahHCqu65zcPtRXwmAXscwOwoaqu75YvYRAK9yRZUVUbk6wA7l3Ae0uSdsBIoVBVhy7WDqvq7iTfTPK4qrodOBa4tXusAs7uni9brH1KkkYzUigkecXW2qvqggXu9zXAhUn2BO4ATmMwlHVxktOB9cBJC3xvSdICjTp89PSh13sx+Ov+RmBBoVBVNwGzW+k6diHvJ0laHKMOH71meLm7+OzDY6lIktSbhU6d/SMGPy2VJO1CRj2n8HF+9hPR3YFfAy4eV1GSpH6Mek7hbUOvNwHrq2rDGOqRJPVopOGjbmK8/2QwU+p+wP3jLEqS1I9R77x2EvAF4GUMfip6fRKnzpakXcyow0d/Djy9qu4FSDID/CuDq5ElSbuIUX99tNvmQOh8+0FsK0laIkY9UrgiyaeAD3XLvwt8cjwlSZL6sr17ND+WwZTWb07yUuDZXdfngAvHXZwkabK2d6TwLuAtAFV1KXApQJIndn2/PdbqJEkTtb3zAsur6uYtG7u2lWOpSJLUm+2Fwr4P0PfQxSxEktS/7YXCXJLf37IxyasY3EZTkrQL2d45hdcDH0tyKj8LgVlgT+Al4yxMkjR5DxgKVXUP8KwkzwWO6Jo/UVVXj70ySdLEjXo/hWuAa8ZciySpZ16VLElqDAVJUmMoSJIaQ0GS1BgKkqTGUJAkNYaCJKkxFCRJjaEgSWoMBUlSYyhIkhpDQZLU9BYKSXZP8h9J/qVbPjTJ9UnWJbkoyZ591SZJ06rPI4XXAbcNLZ8DvLOqHgvcB5zeS1WSNMV6CYUkBwMvAs7rlgM8D7ikW2UtcEIftUnSNOvrSOFdwJ8C/9ctHwB8t6o2dcsbgIO2tmGS1UnmkszNz8+Pv1JJmiITD4UkvwXcW1ULusdzVa2pqtmqmp2ZmVnk6iRpuo1057VFdgzw4iQvBPYCHg68G9g3ybLuaOFg4K4eapOkqTbxI4WqektVHVxVK4GTgaur6lQGt/s8sVttFXDZpGuTpGm3M12ncAbwxiTrGJxjeH/P9UjS1Olj+KipqmuBa7vXdwDP6LMeSZp2O9ORgiSpZ4aCJKkxFCRJjaEgSWoMBUlSYyhIkhpDQZLUGAqSpMZQkCQ1hoIkqTEUJEmNoSBJagwFSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSY2hIElqDAVJUmMoSJIaQ0GS1BgKkqTGUJAkNYaCJKkxFCRJjaEgSWoMBUlSM/FQSHJIkmuS3JrkliSv69r3T3Jlkq91z/tNujZJmnZ9HClsAv6kqg4HjgJeneRw4Ezgqqo6DLiqW5YkTdDEQ6GqNlbVjd3rHwC3AQcBxwNru9XWAidMujZJmna9nlNIshJ4CnA9sLyqNnZddwPLt7HN6iRzSebm5+cnUqckTYveQiHJPsBHgddX1feH+6qqgNradlW1pqpmq2p2ZmZmApVK0vToJRSS7MEgEC6sqku75nuSrOj6VwD39lGbJE2zPn59FOD9wG1V9Y6hrsuBVd3rVcBlk65Nkqbdsh72eQzwcuDmJDd1bX8GnA1cnOR0YD1wUg+1SdJUm3goVNVngWyj+9hJ1iJJ+nle0SxJagwFSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSY2hIElqDAVJUmMoSJIaQ0GS1BgKkqTGUJAkNYaCJKkxFCRJjaEgSWoMBUlSYyhIkhpDQZLUGAqSpMZQkCQ1hoIkqTEUJEmNoSBJagwFSVJjKEiSGkNBktQYCpKkZqcKhSTPT3J7knVJzuy7HkmaNjtNKCTZHXgv8ALgcOCUJIf3W5UkTZedJhSAZwDrquqOqrof+DBwfM81SdJUWdZ3AUMOAr45tLwBeOaWKyVZDazuFn+Y5PYJ1DYOBwLf6ruIJazXzy/n9LXnReV3cMcs5e/go7fVsTOFwkiqag2wpu86dlSSuaqa7buOpcrPb8f5Ge6YXfXz25mGj+4CDhlaPrhrkyRNyM4UCl8EDktyaJI9gZOBy3uuSZKmyk4zfFRVm5L8MfApYHfgA1V1S89ljdOSHwLrmZ/fjvMz3DG75OeXquq7BknSTmJnGj6SJPXMUJAkNYZCD5KckKSSPL7vWpaaJL+c5MNJ/ivJDUk+meRX+65rKUhyQJKbusfdSe4aWt6z7/qWgiTLk/xTkju679/nkryk77oWk6HQj1OAz3bPGlGSAB8Drq2qX6mqpwFvAZb3W9nSUFXfrqojq+pI4FzgnZuXu1kE9AC6798/A5+pqsd037+TGfx8fpdhKExYkn2AZwOnM/hCaXTPBX5aVedubqiqL1XVv/VYk6bH84D7t/j+ra+q9/RY06IzFCbveOCKqvoq8O0kT+u7oCXkCOCGvovQ1HoCcGPfRYyboTB5pzCY7I/u2SEkaQlK8t4kX0ryxb5rWUw7zcVr0yDJ/gwOQZ+YpBhcpFdJ3lxeMDKKW4AT+y5CU+sW4Hc2L1TVq5McCMz1V9Li80hhsk4E/rGqHl1VK6vqEODrwK/3XNdScTXwkG6mXACSPCmJn58m4WpgryR/ONS2d1/FjIuhMFmnMPj1zLCP4hDSSLqjqZcAx3U/Sb0F+Gvg7n4r0zTovn8nAL+Z5OtJvgCsBc7ot7LF5TQXkqTGIwVJUmMoSJIaQ0GS1BgKkqTGUJAkNYaCpk43Q+3bh5bflOStC3yv85Ms6IK6JK9M8sih5fOSHL6N9f5+IfuQHixDQdPof4CXdlej9umVQAuFqnpVVd3aXzmSoaDptInB/XXfsGXHln/5J/nh0OszktzczXdz9la2fVqS67p59j+VZEXXfmSSzyf5cpKPJdmv28cscGF3P4OHJrk2yWy3zWlJvtpdIHXM0D5mknw0yRe7xzFb1iHtCENB0+q9wKlJHjHKyklewGCG22dW1ZOBv9mifw/gPcCJ3Tz7HwD+quu+ADijqp4E3AycVVWXMJgz59TufgY/GXqvFcBfMAiDZwPDQ0rvZnAfhKczmIfnvAf3z5YemBPiaSpV1feTXAC8FvjJ9tYHjgM+WFU/7rb/zhb9j2MwtfeVg3uxsDuwsQudfavqum69tcBHtrOvZzK4kdA8QJKLgM13lzsOOLzbB8DDk+xTVT/8xbeRHjxDQdPsXQzmx//gUNsmuiPoJLsBo96mMsAtVXX0zzWOeCTyIOwGHFVV/73I7ysBDh9pinV/7V/M4C54m90JbL7x0YuBPbrXVwKnJdkb2jTow24HZpIc3fXvkeQJVfU94L6hmVxfDmw+avgB8LCtlHY9g0nXDuiGpV421Pdp4DWbF5IcOeI/VxqJoaBp93Zg+FdI/8DgP+QvAUcDPwKoqiuAy4G5JDcBbxp+k+4exycC53Tb3gQ8q+teBfxtki8DRwJ/2bWfD5y7+UTz0HttBN4KfA74d+C2oV29FpjtTlrfCvzBDv3rpS04S6okqfFIQZLUGAqSpMZQkCQ1hoIkqTEUJEmNoSBJagwFSVLz/4cImOJIyalXAAAAAElFTkSuQmCC\n",
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ]
          },
          "metadata": {
            "needs_background": "light"
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sVa7fDW5ujVK"
      },
      "source": [
        "Lets now see what happens if make the genome 90% GC (45 and 45) and 10% AT (5 and 5)."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 282
        },
        "id": "MytZrC3BugzC",
        "outputId": "b543816a-8cae-40e7-aff0-7bd907682ad5"
      },
      "source": [
        "genome = generate_genome(size = 1000, g_content = 45, c_content = 45, a_content=5) #Make genome size manageable for dynamic plotting.\n",
        "As = 0\n",
        "Ts = 0\n",
        "Gs = 0\n",
        "Cs = 0\n",
        "\n",
        "while (As+Ts+Gs+Cs)<=200:\n",
        "    \n",
        "    clear_output(wait=True)\n",
        "    fig,ax = subplots(ncols= 1,nrows =1, sharey = False)\n",
        "    ax.bar(['A','C','T','G'],[As,Cs,Ts,Gs])\n",
        "    ax.set_xlabel('Nucleotide')\n",
        "    ax.set_ylabel('Count')\n",
        "    sample = np.random.choice(list(genome),1) # making the string genome a list to sample easily\n",
        "\n",
        "    if sample == 'C':\n",
        "        Cs += 1\n",
        "    if sample == 'T':\n",
        "        Ts += 1\n",
        "    if sample == 'G':\n",
        "        Gs += 1\n",
        "    if sample == 'A':\n",
        "        As += 1        \n",
        "    plt.show()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEJCAYAAAB7UTvrAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAP/ElEQVR4nO3de7BdZX3G8e/DJSqicjuTRmA8WKkWUVGPCmJbBf7Q0gpapFBGI4PNtLVeqwXbP7Cd6RTaWrXWgaagxA4FEbGgMFjKrbVj0QOiGBBNkWgYLkfrXVtM++sfe+V1G3PZIdl7nWR/PzOZs9e71t77yZ4zefKutddaqSokSQLYre8AkqTFw1KQJDWWgiSpsRQkSY2lIElqLAVJUjO2UkjygSQPJfni0Nh+Sa5L8pXu577deJL8TZI1Sb6Q5DnjyiVJ2rxxzhQuAl660dhZwPVVdShwfbcM8DLg0O7PCuC8MeaSJG1GxnnyWpJZ4BNVdXi3fDfw4qq6P8ky4KaqemqSv+seX7Lxdlt6/QMOOKBmZ2fHll+SdkW33nrrN6pqZlPr9phwlqVD/9A/ACztHh8IfH1ou3Xd2BZLYXZ2lvn5+R0eUpJ2ZUnWbm5dbweaazBF2eZpSpIVSeaTzC8sLIwhmSRNr0mXwoPdbiO6nw914/cBBw9td1A39jOqamVVzVXV3MzMJmc/kqRHaNKlcBWwvHu8HLhyaPw13beQjgS+s7XjCZKkHW9sxxSSXAK8GDggyTrgbOAc4LIkZwBrgZO7za8BfhVYA/wQOH1cuSRJmze2UqiqUzez6thNbFvA68eVRZI0Gs9oliQ1loIkqbEUJEmNpSBJaiZ9RrN2EbNnXd13hF7de87xfUeQxsKZgiSpsRQkSY2lIElqLAVJUmMpSJIaS0GS1FgKkqTGUpAkNZaCJKmxFCRJjaUgSWosBUlSYylIkhpLQZLUWAqSpMZSkCQ1loIkqbEUJEmNpSBJaiwFSVJjKUiSGktBktRYCpKkxlKQJDWWgiSpsRQkSY2lIElqLAVJUmMpSJKaXkohyVuSrE7yxSSXJHl0kkOS3JJkTZIPJ1nSRzZJmmYTL4UkBwJvBOaq6nBgd+AU4Fzg3VX1FOBbwBmTziZJ026PHt/3MUl+DOwF3A8cA/xWt34V8E7gvF7SSVr0Zs+6uu8Ivbr3nOPH8roTnylU1X3AXwFfY1AG3wFuBb5dVeu7zdYBB046myRNuz52H+0LnAAcAjwReCzw0m14/ook80nmFxYWxpRSkqZTHweajwO+WlULVfVj4ArgaGCfJBt2Zx0E3LepJ1fVyqqaq6q5mZmZySSWpCnRRyl8DTgyyV5JAhwL3AncCJzUbbMcuLKHbJI01fo4pnALcDlwG3BHl2ElcCbw1iRrgP2BCyedTZKmXS/fPqqqs4GzNxq+B3h+D3EkSR3PaJYkNZaCJKmxFCRJjaUgSWosBUlSYylIkhpLQZLUWAqSpMZSkCQ1loIkqbEUJEmNpSBJaiwFSVJjKUiSGktBktRYCpKkxlKQJDWWgiSpsRQkSY2lIElqLAVJUmMpSJIaS0GS1FgKkqTGUpAkNZaCJKmxFCRJjaUgSWosBUlSYylIkhpLQZLUWAqSpMZSkCQ1loIkqemlFJLsk+TyJF9KcleSo5Lsl+S6JF/pfu7bRzZJmmZ9zRTeC1xbVU8DngXcBZwFXF9VhwLXd8uSpAmaeCkkeQLwy8CFAFX1cFV9GzgBWNVttgo4cdLZJGna9TFTOARYAD6Y5HNJLkjyWGBpVd3fbfMAsLSHbJI01foohT2A5wDnVdWzgR+w0a6iqiqgNvXkJCuSzCeZX1hYGHtYSZomfZTCOmBdVd3SLV/OoCQeTLIMoPv50KaeXFUrq2ququZmZmYmEliSpsXES6GqHgC+nuSp3dCxwJ3AVcDybmw5cOWks0nStNujp/d9A3BxkiXAPcDpDArqsiRnAGuBk3vKJklTq5dSqKrbgblNrDp20lkkST/hGc2SpMZSkCQ1loIkqRmpFJIcPcqYJGnnNupM4X0jjkmSdmJb/PZRkqOAFwIzSd46tOrxwO7jDCZJmrytfSV1CbB3t93jhsa/C5w0rlCSpH5ssRSq6mbg5iQXVdXaCWWSJPVk1JPXHpVkJTA7/JyqOmYcoSRJ/Ri1FD4CnA9cAPzv+OJIkvo0aimsr6rzxppEktS7Ub+S+vEkv5dkWXcv5f2S7DfWZJKkiRt1prDhktZvHxor4Mk7No4kqU8jlUJVHTLuIJKk/o1UCkles6nxqvrQjo0jSerTqLuPnjf0+NEM7ntwG2ApSNIuZNTdR28YXk6yD3DpWBJJknrzSC+d/QPA4wyStIsZ9ZjCxxl82wgGF8L7ReCycYWSJPVj1GMKfzX0eD2wtqrWjSGPJKlHI+0+6i6M9yUGV0rdF3h4nKEkSf0Y9c5rJwOfAV4FnAzcksRLZ0vSLmbU3Ud/DDyvqh4CSDID/Atw+biCSZImb9RvH+22oRA639yG50qSdhKjzhSuTfJJ4JJu+TeBa8YTSZLUl63do/kpwNKqenuSVwIv6lZ9Grh43OEkSZO1tZnCe4B3AFTVFcAVAEme0a379bGmkyRN1NaOCyytqjs2HuzGZseSSJLUm62Vwj5bWPeYHRlEktS/rZXCfJLf3ngwyeuAW8cTSZLUl60dU3gz8LEkp/GTEpgDlgCvGGcwSdLkbbEUqupB4IVJXgIc3g1fXVU3jD2ZJGniRr2fwo3AjWPOIknqmWclS5IaS0GS1PRWCkl2T/K5JJ/olg9JckuSNUk+nGRJX9kkaVr1OVN4E3DX0PK5wLur6inAt4AzekklSVOsl1JIchBwPHBBtxzgGH5yKe5VwIl9ZJOkadbXTOE9wB8C/9ct7w98u6rWd8vrgAP7CCZJ02zipZDk14CHquoRnRGdZEWS+STzCwsLOzidJE23PmYKRwMvT3IvcCmD3UbvBfZJsuG8iYOA+zb15KpaWVVzVTU3MzMzibySNDUmXgpV9Y6qOqiqZoFTgBuq6jQGJ8dtuO/zcuDKSWeTpGm3mM5TOBN4a5I1DI4xXNhzHkmaOqPejnMsquom4Kbu8T3A8/vMI0nTbjHNFCRJPbMUJEmNpSBJaiwFSVJjKUiSGktBktRYCpKkxlKQJDWWgiSpsRQkSY2lIElqLAVJUmMpSJIaS0GS1FgKkqTGUpAkNZaCJKmxFCRJjaUgSWosBUlSYylIkhpLQZLUWAqSpMZSkCQ1loIkqbEUJEmNpSBJaiwFSVJjKUiSGktBktRYCpKkxlKQJDWWgiSpsRQkSc3ESyHJwUluTHJnktVJ3tSN75fkuiRf6X7uO+lskjTt+pgprAf+oKoOA44EXp/kMOAs4PqqOhS4vluWJE3QxEuhqu6vqtu6x98D7gIOBE4AVnWbrQJOnHQ2SZp2vR5TSDILPBu4BVhaVfd3qx4AlvYUS5KmVm+lkGRv4KPAm6vqu8PrqqqA2szzViSZTzK/sLAwgaSSND16KYUkezIohIur6opu+MEky7r1y4CHNvXcqlpZVXNVNTczMzOZwJI0Jfr49lGAC4G7quqvh1ZdBSzvHi8Hrpx0Nkmadnv08J5HA68G7khyezf2R8A5wGVJzgDWAif3kE2SptrES6GqPgVkM6uPnWQWSdJP84xmSVJjKUiSGktBktRYCpKkxlKQJDV9fCV1UZg96+q+I/Tq3nOO7zuCpEXImYIkqbEUJEmNpSBJaiwFSVJjKUiSGktBktRYCpKkxlKQJDWWgiSpsRQkSY2lIElqLAVJUmMpSJIaS0GS1FgKkqRmau+nIPVp2u/nAd7TY7FypiBJaiwFSVJjKUiSGktBktRYCpKkxlKQJDWWgiSpsRQkSY2lIElqLAVJUmMpSJIaS0GS1CyqUkjy0iR3J1mT5Ky+80jStFk0pZBkd+D9wMuAw4BTkxzWbypJmi6LphSA5wNrquqeqnoYuBQ4oedMkjRVFlMpHAh8fWh5XTcmSZqQne4mO0lWACu6xe8nubvPPNvhAOAbfb15zu3rnXcYP7/t52e4fXbmz+9Jm1uxmErhPuDgoeWDurGfUlUrgZWTCjUuSearaq7vHDsrP7/t52e4fXbVz28x7T76LHBokkOSLAFOAa7qOZMkTZVFM1OoqvVJfh/4JLA78IGqWt1zLEmaKoumFACq6hrgmr5zTMhOvwusZ35+28/PcPvskp9fqqrvDJKkRWIxHVOQJPXMUuhBkhOTVJKn9Z1lZ5Pk55JcmuQ/k9ya5Jokv9B3rp1Bkv2T3N79eSDJfUPLS/rOtzNIsjTJPya5p/v9+3SSV/Sda0eyFPpxKvCp7qdGlCTAx4Cbqurnq+q5wDuApf0m2zlU1Ter6oiqOgI4H3j3huXuKgLagu7375+Af62qJ3e/f6cw+Pr8LsNSmLAkewMvAs5g8Aul0b0E+HFVnb9hoKo+X1X/1mMmTY9jgIc3+v1bW1Xv6zHTDmcpTN4JwLVV9WXgm0me23egncjhwK19h9DUejpwW98hxs1SmLxTGVzsj+6nu5CknVCS9yf5fJLP9p1lR1pU5yns6pLsx2AK+owkxeAkvUry9vK7waNYDZzUdwhNrdXAb2xYqKrXJzkAmO8v0o7nTGGyTgL+oaqeVFWzVXUw8FXgl3rOtbO4AXhUd1FEAJI8M4mfnybhBuDRSX53aGyvvsKMi6UwWacy+PbMsI/iLqSRdLOpVwDHdV9JXQ38OfBAv8k0DbrfvxOBX0ny1SSfAVYBZ/abbMfyjGZJUuNMQZLUWAqSpMZSkCQ1loIkqbEUJEmNpaCp012h9l1Dy29L8s5H+FoXJXlEJ9QleW2SJw4tX5DksM1s97eP5D2kbWUpaBr9D/DK7mzUPr0WaKVQVa+rqjv7iyNZCppO6xncSvEtG6/Y+H/+Sb4/9PjMJHd017s5ZxPPfW6Sm7vr7H8yybJu/Igk/5HkC0k+lmTf7j3mgIu7+xk8JslNSea655ye5MvdCVJHD73HTJKPJvls9+fojXNI28NS0LR6P3BakieMsnGSlzG4wu0LqupZwF9stH5P4H3ASd119j8A/Fm3+kPAmVX1TOAO4OyqupzBNXNO6+5n8KOh11oG/AmDMngRMLxL6b0M7oPwPAbX4blg2/7a0pZ5QTxNpar6bpIPAW8EfrS17YHjgA9W1Q+75//XRuufyuDS3tcN7sXC7sD9XensU1U3d9utAj6ylfd6AYMbCS0AJPkwsOHucscBh3XvAfD4JHtX1fd/9mWkbWcpaJq9h8H18T84NLaebgadZDdg1NtUBlhdVUf91OCIM5FtsBtwZFX99w5+XQlw95GmWPe//csY3AVvg3uBDTc+ejmwZ/f4OuD0JHtBuwz6sLuBmSRHdev3TPL0qvoO8K2hK7m+Gtgwa/ge8LhNRLuFwUXX9u92S71qaN0/A2/YsJDkiBH/utJILAVNu3cBw99C+nsG/yB/HjgK+AFAVV0LXAXMJ7kdeNvwi3T3OD4JOLd77u3AC7vVy4G/TPIF4AjgT7vxi4DzNxxoHnqt+4F3Ap8G/h24a+it3gjMdQet7wR+Z7v+9tJGvEqqJKlxpiBJaiwFSVJjKUiSGktBktRYCpKkxlKQJDWWgiSpsRQkSc3/A4y5qOb3GYiDAAAAAElFTkSuQmCC\n",
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ]
          },
          "metadata": {
            "needs_background": "light"
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "40JhabZ_uFKX"
      },
      "source": [
        "This example basically shows us that **surprise** is inversely related to probability, the more likely an event is to occur the less surprised we'll be if it actually occurs and vice-versa.\n",
        "\n",
        "In other words and specifically for our example if the probability of sampling an A or an T is low, the **surprise** of sampling either is high. And if the probability of sampling an C or an G is high, the **surprise** of sampling either is low.\n",
        "\n",
        "A good way to describe this mathematically is by taking the logarithm of the inverse of the probability:\n",
        "\n",
        "$$\\mbox{Surprise}= S = \\log{\\frac{1}{p(x)}}$$\n",
        "\n",
        "In that case we could calculate the amount of surpise we'd expect for our genomes, for this last case it would be:\n",
        "\n",
        "\\begin{aligned}\n",
        "S_{A} &= \\log{\\frac{1}{p(A)}} = \\log{\\frac{1}{0.05}} \\approx 3 \\\\\n",
        "S_{T} &= \\log{\\frac{1}{p(T)}} = \\log{\\frac{1}{0.05}} \\approx 3 \\\\\n",
        "S_{C} &= \\log{\\frac{1}{p(C)}} = \\log{\\frac{1}{0.45}} \\approx 0.8 \\\\\n",
        "S_{G} &= \\log{\\frac{1}{p(G)}} = \\log{\\frac{1}{0.45}} \\approx 0.8\n",
        "\\end{aligned}\n",
        "\n",
        "As expected the surprise of sampling a G or a C is smaller than it is for an A or a T.\n",
        "\n",
        "But what about longer motifs? Let's say we want to know the probability of finding the sequence *GGA*, well then the probability of getting an G $p(G) = 0.45$ and the probability of getting an A $p(A) = 0.05$, so the probability of getting the motif is:\n",
        "\n",
        "$$p(GGA) = p(G) \\times p(G)  \\times p(A)  = 0.45 \\times 0.45 \\times 0.05  \\approx 0.01 $$\n",
        "\n",
        "If we want to know exactly how surprising it would be so observe this particular we can plug this probability in our previous expression for surprise:\n",
        "\n",
        "$$S_{GGA} = \\log{\\frac{1}{p(GGA)}} = \\log{\\frac{1}{0.01}} \\approx 4.59$$\n",
        "\n",
        "But more importantly notice that the total surprise for the motif is the sum of individual surprises (because of the properties of logarithms.)\n",
        "\n",
        "$$S_{GGA} = \\log{\\frac{1}{p(G)}} + \\log{\\frac{1}{p(G)}} + \\log{\\frac{1}{p(A)}} = \\log{1} - (log{G} + log{G} + log{A}) \\approx 4.59$$\n",
        "\n",
        "So far we have been talking about surprise and this must feel familiar after this weeks class. That is because **entropy (E)** is the average **surprise** per event. \n",
        "\n",
        "Tying this back to our example, how can we calculate the average entropy of a motif of length $l$? \n",
        "\n",
        "Let's first look at the entropy at a single position, which is one base of the possible A, G, C, T. \n",
        "\n",
        "$$E_\\mbox{position} = \\sum_{base \\in \\{A, G, C, T \\} } p(base) \\times surprise_{base}$$\n",
        "\n",
        "Plugging in what we set as the surprise, and simplifying:\n",
        "\n",
        "$$E_\\mbox{position} = -\\sum_{base \\in \\{A, G, C, T \\} } p(base) \\times \\log p(base)$$\n",
        "Assuming positions are independent of each other, we can use the additivity property of entropy to calculate the entropy of a motif of length $l$:\n",
        "\n",
        "$$E_\\mbox{motif} = l \\times E_\\mbox{position} = -l \\big(\\sum_{base \\in \\{A, G, C, T \\} } p(base) \\times \\log p(base) \\big)$$\n",
        "We've arrived at the formula described by shannon for entropy. What this tells us is the average information content per base position. It turns out there's a very concrete interpretation for this: it measures number of bits in a given message.\n",
        "\n",
        "#### - But how can we address the original question we had in mind? *(see top of the notebook if you've forgotten already.)*\n",
        "\n",
        "Let's try to tackle this question using the concept of **Entropy**. And for simplicity we are going to assume from now on that bases in the genome have equal probabilities of occurring along the genome.\n",
        "\n",
        "The entropy of the location of a motif is going to depend on how many times we can fit the available positions of the genome, for large (or circular) genomes of length $L$ the length of the motif $l$ is usually negligible and we can think of the number of times that we can fit the length of the motif in the genome to be the number of bases in the genome. Let $X$ be a discrete random variable representing the location in a genome of length $L$ (i.e. position 1, 2, ... $L$).\n",
        "\n",
        "An expression for the entropy of the location in the genome can be written as follows:\n",
        "\n",
        "$$E_\\mbox{location} = -\\sum_{i=1}^{L} p(x_i) \\log(p(x_i))\n",
        "$$\n",
        "\n",
        "Let's assume all locations of the genome are equally likely for the motif, so the probability of finding the motif at any $x_i$ is:\n",
        "\n",
        "$$p(x_i) = \\frac{1}{L}\n",
        "$$\n",
        "\n",
        "Plugging this back into the equation we get:\n",
        "\n",
        "$$E_\\mbox{location} = - \\log(1/L)\n",
        "$$\n",
        "\n",
        "$$E_\\mbox{location} = \\log(L)\n",
        "$$\n",
        "\n",
        "We can also calculate the information content of our motif:\n",
        "\n",
        "$$E_\\mbox{motif} = -l \\big(\\sum_{base \\in \\{A, G, C, T \\} } p(base) \\times \\log p(base) \\big)$$\n",
        "\n",
        "Let's assume that A, G, C, T are all equally likely, so we have a $p(A) = p(G) = p(C) = p(T) = 1/4$:\n",
        "\n",
        "\n",
        "$$\\begin{aligned}\n",
        "E_\\mbox{motif} &= -l \\big(\\sum_{base \\in \\{A, G, C, T \\} } \\frac{1}{4} \\times \\log \\frac{1}{4} \\big) \\\\\n",
        "&= -l \\log \\frac{1}{4} \\\\\n",
        "&= l \\log 4 \\\\\n",
        "& = 2l \\\\\n",
        "\\end{aligned}$$\n",
        "\n",
        "Great, now what we want is to find the $l$ so that the entropy of our motif is at least as large as the entropy of the location:\n",
        "\n",
        "$$\\begin{aligned}\n",
        "E_\\mbox{location} &\\leq E_\\mbox{motif} \\\\\n",
        "\\log L &\\leq 2l \\\\\n",
        "\\end{aligned}$$\n",
        "\n",
        "Thus:\n",
        "$$ l \\geq \\frac{1}{2}log(L)\n",
        "$$\n",
        "\n",
        "The sequence has to be at least $\\frac{1}{2}log(L)$ long for it to be in a genome with equal base composition. We can check this computationally! \n",
        "\n",
        "Let's generate genomes of size $10\\times10^6$ which means that the minimum sequence length for it to be uniquely found in a genome of that size must be at least 12 bases long (rounding up!). Let's make 10 genomes to sample from. and pick a couple of length 12 motifs to see how many times they come up in the genomes:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fZUOG9Ib_i_J",
        "outputId": "f900181b-ee4d-4237-cf56-4c14df3fe51e"
      },
      "source": [
        "L = 10e6\n",
        "1/2 * np.log2(L)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "11.626748332105768"
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BrN93lRU9z65",
        "outputId": "f93613f1-3bde-49c2-ac0f-6cd7378592e8"
      },
      "source": [
        "%%time\n",
        "genomes = [generate_genome(size = 10e6, g_content = 25, c_content = 25, a_content=25) for g in range(10)]\n",
        "motifs_16 = [generate_genome(size=16, g_content = 25, c_content = 25, a_content=25) for g in range(1000)]\n",
        "motifs_12 = [generate_genome(size=12, g_content = 25, c_content = 25, a_content=25) for g in range(1000)]\n",
        "motifs_11 = [generate_genome(size=11, g_content = 25, c_content = 25, a_content=25) for g in range(1000)]\n",
        "motifs_9 = [generate_genome(size=9, g_content = 25, c_content = 25, a_content=25) for g in range(1000)]"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CPU times: user 25.3 s, sys: 5.68 s, total: 31 s\n",
            "Wall time: 30.9 s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Al8a_RWxIpDo",
        "outputId": "42342d5a-707b-43d5-8ede-cda9c0a08c4c"
      },
      "source": [
        "counts_16 = [[genome.count(motif) for motif in np.random.choice(motifs_16,100)] for genome in genomes]\n",
        "print('100 motifs of length 16 appear, on average ' + str(mean(counts_16)) + u\" \\u00B1 \" + str(round(std(counts_16),2)))\n",
        "\n",
        "counts_12 = [[genome.count(motif) for motif in np.random.choice(motifs_12,100)] for genome in genomes]\n",
        "print('100 motifs of length 12 appear, on average ' + str(mean(counts_12)) + u\" \\u00B1 \" + str(round(std(counts_12),2)))\n",
        "\n",
        "counts_11 = [[genome.count(motif) for motif in np.random.choice(motifs_11,100)] for genome in genomes]\n",
        "print('100 motifs of length 11 appear, on average ' + str(mean(counts_11)) + u\" \\u00B1 \" + str(round(std(counts_11),2)))\n",
        "\n",
        "counts_9 = [[genome.count(motif) for motif in np.random.choice(motifs_9,100)] for genome in genomes]\n",
        "print('100 motifs of length 9 appear, on average ' + str(mean(counts_9)) + u\" \\u00B1 \" + str(round(std(counts_9),2)))\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "100 motifs of length 16 appear, on average 0.0 ± 0.0\n",
            "100 motifs of length 12 appear, on average 0.586 ± 0.76\n",
            "100 motifs of length 11 appear, on average 2.473 ± 1.56\n",
            "100 motifs of length 9 appear, on average 37.99 ± 6.28\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cElGzekfBr1J"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}